### PR TITLE
Tokenizer / Parser for CSS3 syntax

### PR DESCRIPTION
**Pull request by jaytaph**
This is the initial PR for the css3 parser and tokenizer. We are using the tokenizer from @neuodev , but the parser has been rewritten to get all the syntax working correctly.

This setup is basically a rewrite of the csstree (https://github.com/csstree/csstree) in Rust. This is chosen for the simplicity and readability of the parser. It should be easily extendible with other modules if needed at a later stage. However, not all css syntax is parseable by csstree, and thus we need to implement some additional code to make this work

Not everything is rust-idiomatic, but where possible it is rewritten as such.

Note that this is NOT A COMPLETE IMPLEMENTATION. There are still a lot of things missing and has todo!() placeholders. Most standard css3 can be compiled into an AST though.

The next action plans are to add a few issues concerning implementation and testing this system.

### CODE REVIEW

File: src/bin/css3-parser.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+use simple_logger::SimpleLogger;
+use std::fs;
+use std::process::exit;
+
+fn bail(message: &str) -> ! {
+    println!("{message}");
+    exit(1);
+}
+
+fn main() -> Result<()> {
```
Comment (jaytaph) [inline src/bin/css3-parser.rs:16]: This binary allows you to simply parse, and/or tokenize css tree's found online or on file.

File: src/bin/css3-parser.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+use gosub_engine::css3::parser_config::ParserConfig;
+use gosub_engine::css3::{Css3, Error};
+use simple_logger::SimpleLogger;
+use std::fs;
+use std::process::exit;
+
+fn bail(message: &str) -> ! {
+    println!("{message}");
+    exit(1);
+}
```
Comment (emwalker) [inline src/bin/css3-parser.rs:14]: I believe `anyhow` provides a [macro called `bail!`](https://docs.rs/anyhow/latest/anyhow/macro.bail.html) which does the same thing.

File: src/byte_stream.rs [hunk 3]
```diff
@@ -0,0 +1,579 @@
+use crate::html5::tokenizer::{CHAR_CR, CHAR_LF};
+use std::io::Read;
+use std::{fmt, io};
+
+/// Encoding defines the way the buffer stream is read, as what defines a "character".
```
Comment (jaytaph) [inline src/byte_stream.rs:5]: The bytestream is a slightly improved version of the regular "bytes.rs" iterator.

File: src/css3/parser_config.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+use crate::css3::location::Location;
+
+/// Context defines how the data needs to be parsed
+pub enum Context {
```
Comment (jaytaph) [inline src/css3/parser_config.rs:4]: A parser config defines how a parser should behave. What is the context in which to parse the bytestream (is it a stylesheet, is it inline data etc), and we can decide what to do with errors.

File: src/css3/walker.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
+use crate::css3::node::{Node, NodeType};
```
Comment (jaytaph) [inline src/css3/walker.rs:1]: The walker is just an easy way of displaying the tree. It probably has no use outside debugging purposes.

**Review by emwalker — APPROVED** [review]
LGTM.

It's a little unusual to up break the implementation of a struct across a number of different files that aren't their own modules (e.g., any file where you basically have `impl Css3<'_.>`).  But it seems fine.

File: src/css3/parser/block.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+                        // // check for either semicolon, eof, or rcurly
+                        // let t = self.tokenizer.lookahead_sc(0);
+                        // if t.token_type == TokenType::Semicolon {
+                        //     self.consume(TokenType::Semicolon)?;
+                        //     semicolon_seperated = true;
+                        // } else if t.token_type == TokenType::RCurly {
+                        //     self.tokenizer.reconsume();
+                        //     break;
+                        // }
```
Comment (Sharktheone) [inline src/css3/parser/block.rs:121]: Dead code

File: src/css3/parser/calc.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        // test!(parse_calc, "calc(50px + 5 * (100vw - 300px) / (1100 - 300)), 1fr)", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+        // test!(parse_calc, "calc(10px + 20 * (100vw - 300px) / (1100 - 300))", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+        // test!(parse_calc, "calc(20px + (50px - 20px) * ((100vw - 600px) / (1000 - 600)))", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+        // test!(parse_calc, "calc(100px + (200px - 100px) * ((100vh - 500px) / (800 - 500))", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+        // test!(parse_calc, "calc(40% + calc(82vw / 53em))", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+        // test!(parse_calc, "calc(52% / 48px)", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+        // test!(parse_calc, "calc((100% - var(--maxContainerWidth)) / 2)", Box::new(NodeType::AnPlusB { a: "1".to_string(), b: "2".to_string() } ));
+    }
+}
```
Comment (Sharktheone) [inline src/css3/parser/calc.rs:62]: Dead test

File: src/css3/parser/declaration.rs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            TokenType::Delim('*') => { // next
+            }
+            TokenType::Delim('$') => { // next
+            }
+            TokenType::Delim('+') => { // next
+            }
+            TokenType::Delim('#') => { // next
+            }
+            TokenType::Delim('&') => { // next
+            }
```
Comment (Sharktheone) [inline src/css3/parser/declaration.rs:33]:
```suggestion
            TokenType::Delim('*') | TokenType::Delim('$') | TokenType::Delim('+') |
            TokenType::Delim('#') | TokenType::Delim('&') => {} //next
```

File: src/css3/parser/operator.rs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    pub fn parse_operator(&mut self) -> Result<Node, Error> {
+        log::trace!("parse_operator");
+
+        let loc = self.tokenizer.current_location().clone();
+
+        let operator = self.consume_any()?;
+        if let TokenType::Delim(c) = operator.token_type {
+            if ['/', '*', ',', ':', '+', '-', '='].contains(&c) {
+                return Ok(Node::new(NodeType::Operator(c.to_string()), loc));
+            }
```
Comment (Sharktheone) [inline src/css3/parser/operator.rs:15]: This should be a bit more performant

```suggestion
            match &c {
                '/' | '*' | ',' | ':' | '+' | '-' | '=' => {
                    return Ok(Node::new(NodeType::Operator(c.to_string()), loc));
                },
                _ => {},
            }
            
```

File: src/css3/parser/selector.rs [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+
+        let c = self.consume_any_delim()?;
+        if !['=', '~', '|', '^', '$', '*'].contains(&c) {
+            self.tokenizer.reconsume();
+
+            return Err(Error::new(
+                format!("Expected attribute operator, got {:?}", c),
+                loc,
+            ));
+        }
```
Comment (Sharktheone) [inline src/css3/parser/selector.rs:20]: here again

```suggestion
        match &c {
             '=' | '~' | '|' | '^' | '$' | '*' => {
                self.tokenizer.reconsume();
    
                return Err(Error::new(
                    format!("Expected attribute operator, got {:?}", c),
                    loc,
                ));
            },
            _ => {}
        }
```

File: src/css3/parser/selector.rs [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+        //
+        //     // name is already in T
+        //
+        //     if let TokenType::Function(name) = t.token_type {
+        //         let name = name.to_lowercase().as_str();
+        //     } else {
+        //         return Err(Error::new(format!("Unexpected token {:?}", t), self.tokenizer.current_location().clone()));
+        //     }
+        //
+        //     self.consume(TokenType::RParen)?;
```
Comment (Sharktheone) [inline src/css3/parser/selector.rs:206]: dead code

**Review by Sharktheone — CHANGES_REQUESTED** [review]

File: src/bin/css3-parser.rs (hunk 2 above)
Comment (jaytaph) [reply src/bin/css3-parser.rs:14]: added this

File: src/css3/parser/calc.rs (hunk 7 above)
Comment (jaytaph) [reply src/css3/parser/calc.rs:62]: I leave it here for now as these needs to be implemented. The test cases are ok to keep (they cover most syntax)

**Review by Sharktheone — APPROVED** [review]
LGTM

**Pull request merged**